### PR TITLE
bug:  enable `jobs.index` page to update in real-time

### LINKED ISSUE #10555: ui: Jobs list doesn’t show a new job (on a locally-built binary?)
### Nomad version

```
Nomad v1.1.0-dev (b890a1f7eb298d25fcb39fd2adbc9fe3d8ee14c2+CHANGES)
```

### Operating system and Environment details

macOS 10.14.6

### Issue

When viewing the Nomad UI at `localhost:4646` on a locally-built binary, the jobs list doesn’t show a job newly submitted elsewhere. It’s not happening with a released binary 🤔

### Reproduction steps

Build a Nomad binary that includes the UI with `make dev-ui`, run it, open the jobs list at `localhost:4646/ui/jobs`, submit a job via CLI.

#### Expected Result

The job should appear in the jobs list.

#### Actual Result

The job only appears upon refresh, or by leaving and returning to the jobs list.

jobs-no-live-reload

### PR DESCRIPTION
**Pull request by ChaiWithJai**
Resolves #10555

The `jobs.index` view is bound the model set on the `jobs.index` route. This model does not reactively update to changes in the `ember-data` store (instead it is subscribed to a set of `RecordIdenitfiers`... I believe).

That's why we're able to see job updates, but when a new job is added, our view does not update.

To fix the problem, we need to send a notification to the controller after the `query` and associated callback resolves meaning that our payload has successfully been added/updated to our `ember-date` store.

We also have one more challenge handling the wildcard (`*`) namespace which is a defacto `findAll` query. We have updated our `matchingRecord` algorithm to handle this case.

### CODE REVIEW

File: ui/app/adapters/watchable.js [hunk 1]
```diff
@@ -111,33 +111,36 @@ export default class Watchable extends ApplicationAdapter {
...
+          attr: adapter.queryParamsToAttrs[key],
+        }));
+
+        // Remove existing records that match this query. This way if server-side
+        // deletes have occurred, the store won't have stale records.
+        const matchingRecords = store.peekAll(type.modelName).filter((record) =>
+          queryParamsToAttrs.some((mapping) => {
+            if (mapping.queryParam === 'namespace' && query.namespace === '*')
+              return true;
+            return get(record, mapping.attr) === query[mapping.queryParam];
```
Comment (ChaiWithJai) [inline ui/app/adapters/watchable.js:133]: @philrenaud This algorithm needs more work. I tried to make it minimal change by only adding one life, which works great if no filters are applied... however, if we apply the `namespace` filter, we're could run into issues.

File: ui/app/controllers/jobs/index.js [hunk 2]
```diff
@@ -187,14 +189,26 @@ export default class IndexController extends Controller.extend(
...
-    if (!this.model || !this.model.jobs) return [];
-    return this.model.jobs
+    let jobs = null;
+    if (this.currentCount) {
+      jobs = this.store.peekAll('job');
+    } else {
+      jobs = this.model.jobs;
+    }
+
+    if ((!this.model || !this.model.jobs) && jobs) return [];
```
Comment (ChaiWithJai) [inline ui/app/controllers/jobs/index.js:210]: `bug`: This should say if 

```
if (!jobs) return []
```

File: ui/app/services/watch-list.js [hunk 3]
```diff
@@ -1,11 +1,16 @@
...
+import { computed, set } from '@ember/object';
 import { readOnly } from '@ember/object/computed';
+import { inject as service } from '@ember/service';
 import { copy } from 'ember-copy';
 import Service from '@ember/service';
 
 let list = {};
 
 export default class WatchListService extends Service {
+  @service store;
```
Comment (ChaiWithJai) [inline ui/app/services/watch-list.js:10]: `nit`: remove unused dependency

File: ui/app/services/watch-list.js [hunk 4]
```diff
@@ -1,11 +1,16 @@
...
+
+  jobUpdateCount = 0;
```
Comment (ChaiWithJai) [inline ui/app/services/watch-list.js:12]: @philrenaud I tried to make the minimal surface change here. We could try to make this more generic by converting this into a dictionary, however, I imagine we could gut the watcher system once its time for a redesign.

**Review by philrenaud — CHANGES_REQUESTED** [review]
I am still not sure this is working as intended. I am using a management token with 2 jobs running (shown in the return here) on the default namespace, and I see no jobs in my index. (on the main branch, I see both jobs) 
image

**Comment by lgfa29** [comment]
For me it never shows the job, even after a refresh. Here's a replay:
https://app.replay.io/recording/jobs-nomad--8a2cadb9-0747-4eb6-b7e9-a07244a22fdd

**Comment by ChaiWithJai** [comment]
> I am still not sure this is working as intended. I am using a management token with 2 jobs running (shown in the return here) on the default namespace, and I see no jobs in my index. (on the main branch, I see both jobs) image

Looks like a namespace issue. We have this wonky behavior where the namespace filter only appears when there are multiple namespaces configured. I didn't handle the case where a user isn't using namespaces. Good catch!

**Comment by philrenaud** [comment]
Resolved in https://github.com/hashicorp/nomad/pull/20452

**Pull request closed without merging**